src/zarr.cpp

```diff
@@ -770,6 +778,133 @@ zarr::Zarr::recover_threads_()
     }
 }
 
+std::string
+zarr::Zarr::get_chunk_dir_prefix_() const
+{
+    return "";
+}
+
+zarr::ZarrV3::ZarrV3(CompressionParams&& compression_params)
+  : Zarr(std::move(compression_params))
+{
+}
+
+void
+zarr::ZarrV3::write_array_metadata_(
+  size_t level,
+  const ImageShape& image_shape,
+  const acquire::sink::zarr::TileShape& tile_shape) const
+{
+    namespace fs = std::filesystem;
+    using json = nlohmann::json;
+
+    if (!writers_.contains(level)) {
+        return;
+    }
+
+    const uint64_t frame_count = writers_.at(level).front()->frames_written();
+    const auto frames_per_chunk =
+      std::min(frame_count,
+               (uint64_t)get_tiles_per_chunk(
+                 image_shape, tile_shape, max_bytes_per_chunk_));
+
+    json metadata;
+    metadata["attributes"] = json::object();
+    metadata["chunk_grid"] = json::object({
+      { "chunk_shape",
+        json::array({
+          frames_per_chunk,
+          1,
+          tile_shape.height,
+          tile_shape.width,
+        }) },
+      { "separator", std::string(1, dimension_separator_) },
+      { "type", "regular" },
+    });
+    metadata["chunk_memory_layout"] = "C";
+    metadata["data_type"] = sample_type_to_dtype(image_shape.type);
+    metadata["extensions"] = json::array();
+    metadata["fill_value"] = 0;
+    metadata["shape"] = json::array({
+      frame_count,
+      image_shape.dims.channels,
+      image_shape.dims.height,
+      image_shape.dims.width,
+    });
+
+    if (compression_params_.has_value()) {
+        auto params = compression_params_.value();
+        metadata["compressor"] = {
+            { "codec", "https://purl.org/zarr/spec/codec/blosc/1.0" },
+            { "configuration",
+              { { "blocksize", 0 },
+                { "clevel", params.clevel_ },
+                { "cname", params.codec_id_ },
+                { "shuffle", params.shuffle_ } } },
+        };
+    }
+
+    auto path = (fs::path(data_root_) / "meta" / "root" /
+                 (std::to_string(level) + ".array.json"))
+                  .string();
+    write_string(path, metadata.dump(4));
```

Comment:
Optional: not really part of this PR, but I think we should consider passing through the `fs::path` object instead of converting to a `string` here and elsewhere because the first thing we do in `write_string` is convert it back to a path.

Reply:
No, I think this is a fine idea.